subprojects/core/src/main/java/org/gradle/api/internal/initialization/StandaloneDomainObjectContext.java

```diff
@@ -26,18 +28,52 @@
 import java.util.function.Consumer;
 import java.util.function.Function;
 
-public class RootScriptDomainObjectContext implements DomainObjectContext, ModelContainer<Object> {
+public abstract class StandaloneDomainObjectContext implements DomainObjectContext, ModelContainer<Object> {
     private static final Object MODEL = new Object();
-    public static final RootScriptDomainObjectContext INSTANCE = new RootScriptDomainObjectContext();
 
-    public static final RootScriptDomainObjectContext PLUGINS = new RootScriptDomainObjectContext() {
+    /**
+     * A domain object context not tied to any mutable state.
+     *
+     * NOTE: In almost all cases, other than testing, there is a better domain object context to use.
+     */
+    public static final StandaloneDomainObjectContext ANONYMOUS = new StandaloneDomainObjectContext() {
+        @Override
+        public String getDisplayName() {
+            return "unknown";
+        }
+    };
+
+    /**
+     * Domain object context for resolving plugins outside a project's buildscript.
+     */
+    public static final StandaloneDomainObjectContext PLUGINS = new StandaloneDomainObjectContext() {
         @Override
         public boolean isPluginContext() {
             return true;
         }
+
+        @Override
+        public String getDisplayName() {
+            return "plugin resolution";
+        }
     };
 
-    private RootScriptDomainObjectContext() {
+    /**
+     * A domain object context for resolution within some script.
+     *
+     * TODO: We should probably have more specialized types for each script, to ensure locks
+     * on mutable state are properly acquired, build ID is properly reported, etc.
+     */
+    public static StandaloneDomainObjectContext forScript(ScriptSource source) {
```

Comment:
💅 `final`

Reply:
When I add this, the IDE warns that we are setting a `static` method as `final`. 

In practice I tend to agree final doesn't make much sense for a static method

Reply:
I suppose this is purely a style choice with downsides either way.  Sigh, we need sealed class hierarchies badly.